kimchi_state/src/navigation_manager.cpp

```diff
@@ -133,16 +90,46 @@ void NavigationManager::startLocating(const Point2D& point) {
   localize_goal.pose_estimate.x = point.x;
   localize_goal.pose_estimate.y = point.y;
 
-  auto localize_options =
-      rclcpp_action::Client<GlobalLocalization>::SendGoalOptions();
-  localize_options.goal_response_callback =
-      std::bind(&NavigationManager::localizeGoalResponseCallback, this, _1);
-  localize_options.feedback_callback =
-      std::bind(&NavigationManager::localizeFeedbackCallback, this, _1, _2);
-  localize_options.result_callback =
-      std::bind(&NavigationManager::localizeResultCallback, this, _1);
-  global_localization_action_client_ptr_->async_send_goal(localize_goal,
-                                                          localize_options);
+  auto goal_handle_future = global_localization_action_client_ptr_->async_send_goal(localize_goal);
+  if (goal_handle_future.wait_for(std::chrono::seconds(5)) != std::future_status::ready)
+  {
+    RCLCPP_ERROR(node_->get_logger(), "[NavigationManager] LocalizationGoal was not accepted by server in time.");
+    mission_observer_->onLocalizationCancelled();
+  }
+
+  auto goal_handle = goal_handle_future.get();
+  if (!goal_handle) {
+    RCLCPP_ERROR(node_->get_logger(), "[NavigationManager] LocalizationGoal was rejected by server.");
+    mission_observer_->onLocalizationCancelled();
+  }
+
+  auto result_future = global_localization_action_client_ptr_->async_get_result(goal_handle);
+  mission_observer_->onLocalizationStarted();
+  RCLCPP_INFO(node_->get_logger(), "[Worker Thread] Goal accepted. Waiting for result for up to 60 seconds...");
+
+  // Block and wait for the result for a maximum of 60 seconds.
+  std::future_status status = result_future.wait_for(std::chrono::seconds(60));
```

Comment:
Question: Why are you setting a timer in the client instead of adding a timeout to the action server?

Reply:
When I tried this approach the hole process got stuck inside the action and there was no way to cancel it, even from the terminal sending a cancel signal.